### PR TITLE
Make RegExp type inherently immutable

### LINKED ISSUE #41834: [Bug]: The compiler does not allow intersection between `string:RegExp` and `readonly`
### Description

The following type definition is not allowed, as it generates the error message: `invalid intersection type 'string:RegExp & readonly': no intersection(BCE2026)`.

```ballerina
type firstType string:RegExp & readonly;
```

Although the type of `string:RegExp` is `any`, the error message is not generated if we replace `string:RegExp` with its original type.

```ballerina
type secondType any & readonly;
```

### Steps to Reproduce

Compile the following Ballerina code:

```ballerina
type firstType string:RegExp & readonly;
```

### Affected Version(s)

Ballerina 2201.8.3 (Swan Lake Update 8)

### OS, DB, other environment details and versions

OS: macOS 14.1.1 23B81
JDK: openjdk version "17.0.8" 2023-07-18

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by LakshanWeerasinghe**
## Purpose
$subject

Fixes #41834

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 1]
```diff
@@ -1498,6 +1498,7 @@ public boolean isInherentlyImmutableType(BType type) {
             case TypeTags.INVOKABLE:
             case TypeTags.TYPEDESC:
             case TypeTags.HANDLE:
+            case TypeTags.REGEXP:
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:1501]: Need to update the runtime type checker also. Please check how we can add tests for that too.

File: tests/jballerina-unit-test/src/test/resources/test-src/types/regexp/regexp_type_test.bal [hunk 2]
```diff
@@ -105,6 +106,9 @@ function testRegExpWithUserDefinedType() {
     assertEquality(re `AB*|[^c-d]{1,5}`, x7);
 }
 
+type T1 string:RegExp & readonly;
+type T2 regexp:RegExp & readonly;
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/regexp/regexp_type_test.bal:110]: Shall we add a test for a local variable of such an intersection type too. Group it with this and use an appropriate name to indicate what we are testing.

File: tests/jballerina-unit-test/src/test/resources/test-src/types/regexp/regexp_type_test.bal [hunk 3]
```diff
@@ -105,6 +106,9 @@ function testRegExpWithUserDefinedType() {
...
+
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/regexp/regexp_type_test.bal:111]: Shall we also add tests for 

```ballerina
import ballerina/lang.regexp;

type Foo record {|
    int e;
    readonly regexp:RegExp f;
|};

Foo & readonly rf = {e: 1, f: re `test`};
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java (hunk 1 above)
Comment (LakshanWeerasinghe) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:1501]: Updated in [678ba28](https://github.com/ballerina-platform/ballerina-lang/pull/41843/commits/678ba2815e15240bb6accfae597acec00e0e5455)

File: tests/jballerina-unit-test/src/test/resources/test-src/types/regexp/regexp_type_test.bal (hunk 2 above)
Comment (LakshanWeerasinghe) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/regexp/regexp_type_test.bal:110]: Updated in [dcf7af](https://github.com/ballerina-platform/ballerina-lang/pull/41843/commits/dcf7afbfeff44123692ed2e30c555553bdba2ec7)

**Review by KavinduZoysa — DISMISSED** [review]

**Comment by KavinduZoysa** [comment]
@LakshanWeerasinghe, please sync this with the master.

File: tests/jballerina-unit-test/src/test/resources/test-src/types/readonly/test_inherently_immutable_type.bal [hunk 4]
```diff
@@ -163,6 +163,12 @@ function testRuntimeIsTypeForInherentlyImmutableBasicTypes() {
     'xml:Text xmlText = xml `xml text`;
     any n = xmlText;
     assertTrue(n is readonly);
+    
+    any reg1 = re `pattern`;
+    assertTrue(reg1 is readonly);
+    
+    readonly reg2 = re `pattern`;
+    assertTrue(reg2 is readonly);
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/readonly/test_inherently_immutable_type.bal:171]: I would add a test with something like `regexp:fromString(str)` where the value is created at runtime with no compile-time input.

**Review by KavinduZoysa — APPROVED** [review]

File: tests/jballerina-unit-test/src/test/resources/test-src/types/regexp/regexp_type_test.bal [hunk 5]
```diff
@@ -105,6 +106,22 @@ function testRegExpWithUserDefinedType() {
...
+    int e;
+    readonly regexp:RegExp f;
+|};
+
+Foo & readonly rf = {e: 1, f: re `test`};
+
+function testRegExpReadonlyLocalVars() {
+    string:RegExp & readonly x1 = re `test`;
+    T1 & readonly x2 = re `test`;
+    (T2 & readonly) & string:RegExp x3 = re `test`;
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/regexp/regexp_type_test.bal:122]: Can use _ instead of the variable names.

**Review by MaryamZi — APPROVED** [review]

File: tests/jballerina-unit-test/src/test/resources/test-src/types/regexp/regexp_type_test.bal (hunk 5 above)
Comment (LakshanWeerasinghe) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/regexp/regexp_type_test.bal]: Updated in commit [#6c0d98](https://github.com/ballerina-platform/ballerina-lang/pull/41843/commits/6c0d989173879eca2bb63f3fa733c0c7ed2ee69d).

File: tests/jballerina-unit-test/src/test/resources/test-src/types/regexp/regexp_type_test.bal [hunk 6]
```diff
@@ -105,6 +106,22 @@ function testRegExpWithUserDefinedType() {
...
+    readonly regexp:RegExp f;
+|};
+
+Foo & readonly rf = {e: 1, f: re `test`};
+
+function testRegExpReadonlyLocalVars() {
+    string:RegExp & readonly _ = re `test`;
+    T1 & readonly _ = re `test`;
+    (T2 & readonly) & string:RegExp _ = re `test`;
+}
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/regexp/regexp_type_test.bal:123]: Don't we need assertions here? Also kinda duplicating the other test, right?

File: tests/jballerina-unit-test/src/test/resources/test-src/types/regexp/regexp_type_test.bal [hunk 7]
```diff
@@ -105,6 +106,22 @@ function testRegExpWithUserDefinedType() {
...
+
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/regexp/regexp_type_test.bal:124]: Let's add some negative tests too. E.g., trying to use it against other immutable types that are incompatible.

Probably also something like

```
(stream<int>|regexp:RegExp) & readonly x = new;
```

File: tests/jballerina-unit-test/src/test/resources/test-src/types/regexp/regexp_type_test.bal (hunk 6 above)
Comment (LakshanWeerasinghe) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/regexp/regexp_type_test.bal]: Updated in [d4f514](https://github.com/ballerina-platform/ballerina-lang/pull/41843/commits/d4f514b63dce28a4904ac3e21cc278607893631f).

File: tests/jballerina-unit-test/src/test/resources/test-src/types/regexp/regexp_type_test.bal (hunk 7 above)
Comment (LakshanWeerasinghe) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/regexp/regexp_type_test.bal:129]: Added in [0ae6e3](https://github.com/ballerina-platform/ballerina-lang/pull/41843/commits/0ae6e361aead9db6235956d2a6dd141a62ca8396).

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/intersection/IntersectionTypeTest.java [hunk 8]
```diff
@@ -86,6 +86,8 @@ public void testImmutableTypesNegative() {
         validateError(result, index++, "invalid intersection type '(Baz & readonly)': no intersection", 32,
                       45);
         validateError(result, index++, "incompatible types: 'Y' is not a record", 42, 6);
+        validateError(result, index++, "incompatible types: expected " +
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/intersection/IntersectionTypeTest.java:89]: May make sense to add this test in org.ballerinalang.test.types.readonly.SelectivelyImmutableTypeTest instead.

File: tests/jballerina-unit-test/src/test/resources/test-src/types/intersection/test_intersection_type_negative.bal [hunk 9]
```diff
@@ -43,3 +43,9 @@ type Z record {
     X x;
     Y y;
 };
+
+function testUnsupportedIntersectionWithReadonly(stream<int> strm) {
+    (stream<int>|string:RegExp) & readonly x = strm;
+}
+
+
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/intersection/test_intersection_type_negative.bal:51]: Extra new lines.

File: tests/jballerina-unit-test/src/test/resources/test-src/types/intersection/test_intersection_type_negative.bal [hunk 10]
```diff
@@ -43,3 +43,9 @@ type Z record {
     X x;
     Y y;
 };
+
+function testUnsupportedIntersectionWithReadonly(stream<int> strm) {
+    (stream<int>|string:RegExp) & readonly x = strm;
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/intersection/test_intersection_type_negative.bal:48]: Use _ to avoid unused variables.

Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/intersection/test_intersection_type_negative.bal:48]: Can add a positive test with a type like this too.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/intersection/IntersectionTypeTest.java (hunk 8 above)
Comment (LakshanWeerasinghe) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/intersection/IntersectionTypeTest.java]: Updated in [commit](https://github.com/ballerina-platform/ballerina-lang/pull/41843/commits/6f8cd3857971872ba1c12e95f60ebb0d4709452b).

File: tests/jballerina-unit-test/src/test/resources/test-src/types/intersection/test_intersection_type_negative.bal (hunk 9 above)
Comment (LakshanWeerasinghe) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/intersection/test_intersection_type_negative.bal]: Updated in [commit](https://github.com/ballerina-platform/ballerina-lang/pull/41843/commits/6f8cd3857971872ba1c12e95f60ebb0d4709452b).

File: tests/jballerina-unit-test/src/test/resources/test-src/types/intersection/test_intersection_type_negative.bal (hunk 10 above)
Comment (LakshanWeerasinghe) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/intersection/test_intersection_type_negative.bal]: Updated in [commit](https://github.com/ballerina-platform/ballerina-lang/pull/41843/commits/6f8cd3857971872ba1c12e95f60ebb0d4709452b).

Comment (LakshanWeerasinghe) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/intersection/test_intersection_type_negative.bal]: Updated in [commit](https://github.com/ballerina-platform/ballerina-lang/pull/41843/commits/6f8cd3857971872ba1c12e95f60ebb0d4709452b).

File: tests/jballerina-unit-test/src/test/resources/test-src/types/intersection/test_intersection_type_negative.bal [hunk 11]
```diff
@@ -43,3 +43,4 @@ type Z record {
     X x;
     Y y;
 };
+
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/intersection/test_intersection_type_negative.bal:46]: Unnecessary change.

**Review by MaryamZi — APPROVED** [review]

**Pull request merged**